# Importing Libraries

In [2]:
# step 1. Download the data file containing your project data sets for this Exercise, extract it, and move the data files into your “Original Data” folder.
# step 2. Import the “orders.csv” data set into your notebook using the os library and the vars_list shortcut, omitting the “eval_set” column in your final dataframe.
# step 3. Import the pandas, NumPy, and os libraries.
import pandas as pd
import numpy as np
import os

In [3]:
# Turn your project folder path into a string as outlined in the Exercise text.
path = r'/Users/peterreadman/Desktop/Python Projects/CareerFoundry/04-2020 Instacart Basket Analysis/'

In [4]:
# import products.csv data set
df_prods = pd.read_csv(os.path.join(path, '02 Data', '02a Original Data', 'products.csv'), index_col = False)

In [5]:
# import orders_wrangled.csv data set
df_ords = pd.read_csv(os.path.join(path, '02 Data', '02b Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Mixed-Type Data

In [8]:
# Create a dataframe

df_test = pd.DataFrame()

In [9]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
# Update data type of mix to string

df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [13]:
# isnull() assigned to df_prods dataframe, (boolean)
# then sum the result (true =1 false =0), returning count of null rows

In [14]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
# create a subset of missing product_name values

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
# call the new datafram to reveal the missing values

df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
# get the rows and columns of df_products

df_prods.shape

(49693, 5)

In [18]:
# create a subset of non-missing product_name values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [20]:
#Finding duplicates

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
# check the current number of rows in df_prods_clean to compare to

df_prods_clean.shape

(49677, 5)

In [23]:
#create a new dataframe that doesn't include duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

# Tidying Up and Exporting Changes

In [25]:
# export the new clean dataframe

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', '02b Prepared Data', 'products_checked.csv'))

In [7]:
# import products.csv data set
df_prods = pd.read_csv(os.path.join(path, '02 Data', '02a Original Data', 'products_checked.csv'), index_col = False)

FileNotFoundError: [Errno 2] No such file or directory: '/Users/peterreadman/Desktop/Python Projects/CareerFoundry/04-2020 Instacart Basket Analysis/02 Data/02a Original Data/products_checked.csv'

# Task 4.5 Data Consistency Checks

# Step 1
- consistency checks on **df_prods** dataframe **done**

# Step 2

*Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.*
 - Tip: Keep an eye on min and max values!

In [27]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_the_week  order_hour_of_day  days_since_last_order  
count            3.421083e+06       3.421083e+06           3.214874e+06  
mean             2.776219e+00       1.345202e+01           1.111484e+01  
std              2.046829e+00       4.226088e+00           9.206737e+00  
min              0.000000e+00       0.000000e+00           0.000000e+00  
25%              1.000000e+00       1.000000e+01           4.000000e+00  
50%              3.000000e+00       1.300000e+01           7.000000e+00  
75%              5.000000e+00       1.600000e+01           1.500000e+01  
max              6.000000e+00       2.300000e+01           3.000000e+01

In [28]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_the_week      int64
order_hour_of_day           int64
days_since_last_order     float64
dtype: object

### Answer - comments and observations

- There is an 'Unnamed' column that looks like the index which is unexpected as the index_col was not imported (index_col = False)

- there are far fewer user_id's than order_id's, indicating that users make multiple orders and/or there may be missing user_id values

- days of the week are numbered 0–6 (assuming Monday = 0 and Sunday = 6)

- hours of the day are numbered 0–23 (0 = midnight, 23 = 23:00)

- the maximum order_number is 100, but the mean is only 17


# Step 3

*Check for mixed-type data in your df_ords dataframe.*



In [29]:
# Check for mixed types

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

# Step 4

*If you find mixed-type data, fix it. The column in question should contain observations of a single data type.*

**Answer**
- Running the check for mixed types above returns no columns of mixed data type

# Step 5
*Run a check for missing values in your df_ords dataframe.*

In [30]:
# isnull() assigned to df_ords dataframe, (boolean)
# then sum the result (true =1 false =0), returning count of null rows

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_last_order     206209
dtype: int64

In a markdown cell, report your findings and propose an explanation for any missing values you find.

**answer**
- the 'days_since_last_order' column has 206,209 missing values. This may be due to these being the first orders by these particular customers. 


# Step 6
*Address the missing values using an appropriate method.*

In [31]:
# first I'll investigate further to make sure the NaN values in 'days_since_last_order' are 'first orders'
# if they are all 'first orders' they should be 'order_number' = 1 and have unique user_id
# there should be the same count of user_ids as rows of NaN

# create a subset of missing days_since_last_order values

df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [32]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0                 0   2539329        1             1                       2   
11               11   2168274        2             1                       2   
26               26   1374495        3             1                       1   
39               39   3343014        4             1                       6   
45               45   2717275        5             1                       3   
...             ...       ...      ...           ...                     ...   
3420930     3420930    969311   206205             1                       4   
3420934     3420934   3189322   206206             1                       3   
3421002     3421002   2166133   206207             1                       6   
3421019     3421019   2227043   206208             1                       1   
3421069     3421069   3154581   206209             1                       3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

In [33]:
df_ords_nan.describe()

Unnamed: 0      order_id        user_id  order_number  \
count  2.062090e+05  2.062090e+05  206209.000000      206209.0   
mean   1.712637e+06  1.708462e+06  103105.000000           1.0   
std    9.874816e+05  9.881299e+05   59527.555167           0.0   
min    0.000000e+00  2.000000e+01       1.000000           1.0   
25%    8.579730e+05  8.507300e+05   51553.000000           1.0   
50%    1.717068e+06  1.706246e+06  103105.000000           1.0   
75%    2.570518e+06  2.564292e+06  154657.000000           1.0   
max    3.421069e+06  3.421081e+06  206209.000000           1.0   

       orders_day_of_the_week  order_hour_of_day  days_since_last_order  
count           206209.000000      206209.000000                    0.0  
mean                 2.754118          13.626597                    NaN  
std                  2.076205           4.223769                    NaN  
min                  0.000000           0.000000                    NaN  
25%                  1.000000          11.000000                    NaN  
50%                  3.000000          14.000000                    NaN  
75%                  5.000000          17.000000                    NaN  
max                  6.000000          23.000000                    NaN

In [34]:
df_ords_nan.nunique()

Unnamed: 0                206209
order_id                  206209
user_id                   206209
order_number                   1
orders_day_of_the_week         7
order_hour_of_day             24
days_since_last_order          0
dtype: int64

- there are 206,209 unique user_id values and only 1 'order_number' (all are 'order_number' = 1), so these are the first orders by these customers hence no previous order

**Answer**
- I propose making no changes to these values. Imputing a value to replace NaN could be misleading in any analysis since the question is not applicable to first orders.

# Step 7
- Run a check for duplicate values in your df_ords data.

In [35]:
#Finding duplicates

df_dups = df_ords[df_ords.duplicated()]

In [36]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_the_week, order_hour_of_day, days_since_last_order]
Index: []

**Answer**
- no duplicates found

# Step 8
Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [67]:
# df_prods_checked already exported

# export df_ords

df_ords.to_csv(os.path.join(path, '02 Data', '02b Prepared Data', 'orders_checked.csv'))